### DISCUSSION TITLE
[@types/node] Compiler error: Cannot find module 'undici-types'

### DISCUSSION THREAD

**Discussion by ryze312** [DefinitelyTyped/DefinitelyTyped#67406] [Issues with a @types package]
I am unable to compile Firefox extension using [web-ext](https://github.com/mozilla/web-ext) because of the following error in `@types/node` module

```ts
node_modules/@types/node/globals.d.ts:325:84 - error TS2792: Cannot find module 'undici-types'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option?

325         type _Request = typeof globalThis extends { onmessage: any } ? {} : import("undici-types").Request;
...
repeated > 10 times
```

[a comment by an automated account is not shown]

**Comment by Semigradsky** [comment]
Try to reinstall dependencies.

**Comment by ryze312** [comment]
The issue seems to only occur when building with yarn, building with pnpm works as expected

**Reply by chpio** [reply]
i'm getting the same problem with pnpm now

**Comment by baybal** [comment]
Same issue started appearing in a release from 3 weeks ago

**Comment by simonsarris** [comment]
This definitely seems like a bug with the types, running tsc will yield

```
../node_modules/@types/node/globals.d.ts:377:25 - error TS2792: Cannot find module 'undici-types'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option?
```

In a project that does not need or use nodenext for modules. (or undici or any other dependencies)

**Comment by simonsarris** [comment]
If you are coming across this in your own projects, as a workaround, you can `npm i undici-types` and then add this to the `tsconfig.json`

```ts
{
  "compilerOptions": {
    "paths": {
      // This is solely to stop a bug with @types/node as of 12/15/2023
      "undici-types": [
        "./node_modules/undici-types/index.d.ts"
      ],
    }
  }
}
```

And it will work for now.

**Comment by joakim** [comment]
In my case, the culprit was `@types/bun`. Once I removed that, no issues.

**Comment by rezonant** [comment]
This is only an issue if you are not using `moduleResolution=Node` or one of the other Node-specific variants. However, it is quite easy for this to happen even if you are not yourself using `@types/node`, it is instead installed as part of your development dependencies :-\ In that case setting `moduleResolution` to `Node` may not be desirable.

> you can npm i undici-types and then add this to the tsconfig.json

This has no effect for me.

**Comment by Willem3141** [comment]
I just ran into the same issue. Because of this bug, merely _installing_ `@types/node` caused my build to fail:
```
node_modules/@types/node/http.d.ts(1954,32): error TS2792: Cannot find module 'undici-types'. Did you mean to set the 'moduleResolution' option to 'node', or to add aliases to the 'paths' option?
```
... even though the code being compiled didn't use any Node types.

The suggestion to set `"moduleResolution": "node"` makes the error go away. However my project contains both server-side (Node) and client-side TS code, and obviously I cannot set `"moduleResolution": "node"` for the client-side portion. Because simply having `@types/node` installed is enough to trigger the issue, regardless of if the types are actually used, I'm stuck.

For now I've downgraded `@types/node` to avoid the issue. I guess I could start maintaining two separate `node_modules` for server- and client-side, but that seems like an inconvenient workaround for this bug.

**Reply by Renegade334** [reply]
I realise that this is an old comment, but for anyone else having the same problem, adding `"types": []` to the tsconfig will disable the automatic loading of @types modules by TypeScript. This can be helpful if, for example, one of your dependencies is drawing in @types/node through its own dependency tree, without being needed in your own build.

**Comment by Josef-adilson27** [comment]
Some external libraries may require the use of Commonjs. In such cases, you may encounter problems, and you will need to find alternative solutions .Adding **"Moduleresolution": "Node" to your TSCONFIG.JSON**  file helps the Typescript compiler to determine how to allow modules. This can be useful if you switch your project to use ES6 modules, especially when problems arise with resolving modules and conflicts with various modules systems.

**Comment by el2-d2-42** [comment]
I had the same issue and what resolved it was to create a separate config file for the type script files that were causing it (in my case client.ts):

`{
    "compilerOptions": {
        "target": "ES2018",
        "lib": [
            "dom",
            "es2018"
        ],
        "outDir": "./public/js",
        "module": "none",
        "skipLibCheck": true
    },
    "files": [
        "src/client.ts"
    ]
}`

**Comment by damienstanton** [comment]
For me, this is causing failure on a basic TypeScript project with no dependencies other than `@types/node`. Installing `undici-types` as a dev dependency does not fix the issue.

**Comment by borgboyone** [comment]
I just ran into this nasty issue after installing the selenium-webdriver and chrome-driver.  Adding "types": [], to my tsconfig.json file did remove the errors from my compilation step. However, I knew this didn't address the underlying issue. Taking a look at my "node_modules" directory under my project and snooping around revealed the root cause of the problem.  The fix explains it all.

Go into the "node_modules/@types" folder.  Look for "node/node_modules/undici-types".  Move the "undici-types" folder so that it is under"node_modules/@types".

That did the trick for me.

**Comment by KhuiKhui** [comment]
Just ran into this issue, enabling "skipLibCheck": true skips type-checking node_modules and solved the problem for me.

**Comment by acxz** [comment]
This helped me:

https://github.com/fable-compiler/Fable/issues/3612#issuecomment-1826423289
